src/blas.jl

```diff
@@ -28,44 +12,88 @@ kernel(::Type{Layout.AlignedRowMajor{T}}, ::Type{Layout.AlignedColMajor{T}}) whe
 kernel(::Type{Layout.AlignedRowMajor{T}}, ::Type{Layout.AlignedRowMajor{T}}) where {T} = Kernel.matmul_pipelined
 
 # Based on https://github.com/JuliaGPU/CUDA.jl/blob/bd5a2a8800e91eb6a7df89eb5dd4bb8fc503541d/lib/cublas/wrappers.jl#L743-L769
-function gemmEx!(transA::Char, transB::Char, alpha::Number, A, B, beta::Number, C)
+function gemmEx!(transA::Char, transB::Char, alpha::Number, A::CuMatrix, B::CuMatrix,
+                 beta::Number, C::CuMatrix; wmma::Union{Bool,Nothing}=nothing)
     m = size(A, transA == 'N' ? 1 : 2)
     k = size(A, transA == 'N' ? 2 : 1)
     n = size(B, transB == 'N' ? 2 : 1)
-
     if m != size(C, 1) || n != size(C, 2) || k != size(B, transB == 'N' ? 1 : 2)
         throw(DimensionMismatch("Dimensions do not match"))
     end
 
     transpose_a = (transA == 'T')
     transpose_b = (transB == 'T')
-
-    a_layout = global_layout(typeof(A), Val(transpose_a))
-    b_layout = global_layout(typeof(B), Val(transpose_b))
-
-    conf = GemmKernels.get_config(
+    a_layout_base = transpose_a ? Layout.AlignedRowMajor : Layout.AlignedColMajor
+    b_layout_base = transpose_b ? Layout.AlignedRowMajor : Layout.AlignedColMajor
+
+    # determine global memory layouts
+    ## if alpha is zero, we don't need to load A or B
+    if iszero(alpha)
+        global_a_layout = Layout.Zero{eltype(A)}
+        global_b_layout = Layout.Zero{eltype(B)}
+    else
+        global_a_layout = a_layout_base{eltype(A)}
+        global_b_layout = b_layout_base{eltype(B)}
+    end
+    ## if beta is zero, we don't need to load C
+    global_c_layout = if iszero(beta)
+        Layout.Zero{eltype(C)}
+    else
+        Layout.AlignedColMajor{eltype(C)}
+    end
+    global_d_layout = Layout.AlignedColMajor{eltype(C)}
+
+    # determine shared memory layouts (padded to avoid bank conflicts)
+    compute_type = promote_type(eltype(A), eltype(B))
+    ## padded to avoid bank conflicts, and converting values to the compute type domain
+    shared_a_layout = Layout.Padded{a_layout_base{compute_type}, 8}
+    shared_b_layout = Layout.Padded{b_layout_base{compute_type}, 8}
+    ## outputs are never transposed
+    ## XXX: why not padded?
+    storage_type = eltype(C)
+    shared_c_layout = shared_d_layout = Layout.AlignedColMajor{storage_type}
```

Comment:
FYI, padding on C or D does not seem worth it: the extra padding would mean that the tile sizes for C, and hence A or B must be made smaller, which lowers the reuse factor in shared memory. Since C / D is only loaded / stored once anyway, I opted for no padding.